xwiki-platform-core/xwiki-platform-attachment/xwiki-platform-attachment-api/src/main/java/org/xwiki/attachment/internal/job/MoveAttachmentJob.java

```diff
@@ -0,0 +1,157 @@
+    private EntityReferenceSerializer<String> entityReferenceSerializer;
+
+    @Inject
+    private ContextualLocalizationManager contextualLocalizationManager;
+
+    @Inject
+    private EntityReferenceSerializer<String> referenceSerializer;
+
+    @Inject
+    private AttachmentsManager attachmentsManager;
+
+    @Override
+    public String getType()
+    {
+        return HINT;
+    }
+
+    @Override
+    protected void process(EntityReference source)
+    {
+        this.progressManager.pushLevelProgress(this);
+        this.request.setId("refactoring", "moveAttachment",
+            String.format("%d-%d", System.currentTimeMillis(), ThreadLocalRandom.current().nextInt(100, 1000)));
+
+        AttachmentReference destination = this.request.getProperty(MoveAttachmentRequest.DESTINATION);
+        boolean autoRedirect = this.request.getProperty(MoveAttachmentRequest.AUTO_REDIRECT);
+
+        XWiki wiki = this.xcontextProvider.get().getWiki();
+        try {
+            XWikiDocument sourceDocument = wiki.getDocument(source.getParent(), this.xcontextProvider.get());
+            XWikiDocument targetDocument = wiki.getDocument(destination.getParent(), this.xcontextProvider.get());
+            XWikiAttachment sourceAttachment = sourceDocument.getAttachment(source.getName());
+
+            // Remove the original attachment and create a new one with the same name.
+            sourceDocument.removeAttachment(sourceAttachment);
+
+            targetDocument.setAttachment(destination.getName(),
+                sourceAttachment.getContentInputStream(this.xcontextProvider.get()), this.xcontextProvider.get());
+
+            this.attachmentsManager.removeExistingRedirection(destination.getName(), targetDocument);
+
+            if (autoRedirect) {
+                initializeAutoRedirection(source, destination, sourceDocument);
+            }
+
+            if (Objects.equals(source.getParent(), destination.getParent())) {
+                wiki.saveDocument(sourceDocument,
+                    this.contextualLocalizationManager.getTranslationPlain("attachment.job.saveDocument.inPlace",
+                        source.getName(), destination.getName()),
+                    this.xcontextProvider.get());
+            } else {
+                // "Attachment moved to " + destination
+                String historyMessageSource =
+                    this.contextualLocalizationManager.getTranslationPlain("attachment.job.saveDocument.source",
+                        this.referenceSerializer.serialize(destination));
+                String historyMessageTarget =
+                    this.contextualLocalizationManager.getTranslationPlain("attachment.job.saveDocument.target",
+                        this.referenceSerializer.serialize(source));
+                wiki.saveDocument(sourceDocument, historyMessageSource, this.xcontextProvider.get());
+                wiki.saveDocument(targetDocument, historyMessageTarget, this.xcontextProvider.get());
```

Comment:
So in case of exception here, you have no flag informing that the previous save should be rollback? I remember you asked about that, but I guess you could have put a boolean flag and check it in the catch to rollback the previous save no?

Reply:
Indeed, I'll add a rollback to put the attachment back in the source if the target save fails.